src/sql/SqlResult.ts

```diff
@@ -0,0 +1,380 @@
+            }
+            return new SqlRowImpl(values, this.rowMetadata);
+        } else { // Return objects
+            const result: SqlRowAsObject = {};
+            for (let i = 0; i < this.currentPage.getColumnCount(); i++) {
+                const columnMetadata = this.rowMetadata.getColumnByIndex(i);
+                result[columnMetadata.name] = this.serializationService.toObject(
+                    this.currentPage.getValue(this.currentPosition, i)
+                );
+            }
+            return result;
+        }
+    }
+
+    /** Called when a execute response is received. */
+    onExecuteResponse(rowMetadata: SqlRowMetadata | null, rowPage: SqlPage, updateCount: Long) {
+        // Ignore the response if SQL result is closed.
+        if (this.closed) return;
+
+        if (rowMetadata !== null) { // Result that including rows
+            this.rowMetadata = rowMetadata;
+            this.onNextPage(rowPage);
+            this.updateCount = Long.fromInt(-1);
+        } else { // Result that including update count
+            this.updateCount = updateCount;
+            this.closed = true;
+        }
+        this.executeDeferred.resolve(true);
+    }
+
+    /**
+     * Fetches the next page. Called internally by iteration logic. Pages are fetched whenever the current page is fully
+     * iterated.
+     */
+    fetch(): Promise<SqlPage> {
+        // If there is an ongoing fetch, return that promise
+        if (this.fetchDeferred?.promise)
+            return this.fetchDeferred.promise;
+
+        // Do not start a fetch if close is called
+        if (this.closeDeferred?.promise) {
+            return Promise.reject('Cannot fetch, close() is called');
+        }
+        // Do not start a fetch if the result is already closed
+        if (this.closed) {
+            return Promise.reject('Cannot fetch, the result is already closed');
+        }
+
+        this.fetchDeferred = deferredPromise<SqlPage>();
+
+        this.sqlService.fetch(this.connection, this.queryId, this.cursorBufferSize).then(value => {
+            this.fetchDeferred.resolve(value);
+            this.fetchDeferred = undefined; // Set fetchDeferred to undefined to be able to fetch again
+        }).catch(err => {
+            this.fetchDeferred.reject(this.sqlService.toHazelcastSqlException(err, this.connection));
+        });
+
+        return this.fetchDeferred.promise;
+    }
+
+    /** Returns if there are rows to be iterated. Used by {@link next}. */
+    hasNext(): Promise<boolean> {
+        return this.executeDeferred.promise.then(() => {
+            if (this.updateCount !== Long.fromInt(-1)) {
+                return Promise.reject(new IllegalStateError('This result contains only update count'));
+            }
+            // While loop similar logic is written via recursion when written in async manner
+            const checkHasNext = (): Promise<boolean> => {
```

Comment:
I don't know much about this topic in Node.js but, this closure does not capture anything from above. So, it can be moved into a separate function.

Will this improve performance? Because, from what I see, we are creating a closure each time we call `hasNext`. This cost can be avoided with the approach I mentioned

Reply:
not sure but will do, reasonable